client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -120,10 +125,12 @@ public synchronized void setHosts(HttpHost... hosts) {
         if (hosts == null || hosts.length == 0) {
             throw new IllegalArgumentException("hosts must not be null nor empty");
         }
+        authCache.clear();
         Set<HttpHost> httpHosts = new HashSet<>();
         for (HttpHost host : hosts) {
             Objects.requireNonNull(host, "host cannot be null");
             httpHosts.add(host);
+            authCache.put(host, new BasicScheme());
```

Comment:
This should probably be `volatile` and work in the same way as `hosts`. Built, then set.

I wanted to avoid adding an unnecessary `volatile` variable; `BasicAuthCache` is thread safe, but clearing it technically means that an async request could be sent that then does not use preemptive authorization _while_ the new hosts are being set.